File: leetcode/python/000684.redundant-connection.py
class Solution:
    # Time complexity: O(V + E)
    # Space complexity: O(V)
    def findRedundantConnection(self, edges: list[list[int]]) -> list[int]:
        N = len(edges)
        par = [i for i in range(N + 1)]  # i-th node -> parent
        rank = [1] * (N + 1)

        def find(n):
            if n != par[n]:
                par[n] = find(par[n])
            return par[n]

        def union(n1, n2):
            p1, p2 = find(n1), find(n2)
            if p1 == p2:
                return False

            if rank[p1] > rank[p2]:
                par[p2] = p1
                rank[p1] += rank[p2]
            else:
                par[p1] = p2
                rank[p2] += rank[p1]

            return True

        for n1, n2 in edges:
            if not union(n1, n2):
                return [n1, n2]

        return edges[0]  # won't happen

    def findRedundantConnection2(self, edges: list[list[int]]) -> list[int]:
        N = len(edges)
        parents = [i for i in range(N + 1)]  # i-th node -> parent, ignore 0

        def find_par(x):
            if parents[x] != x:
                parents[x] = find_par(parents[x])  # path compression
            return parents[x]

        def union(x, y):
            root_x = find_par(x)
            root_y = find_par(y)

            if root_x == root_y:
                return False

            parents[root_x] = root_y # make 1 root point to the other (union)
            return True

        for x, y in edges:
            if not union(x, y)
                return [x, y]

        return edges[0]  # won't happen

File: leetcode/javascript/000684.redundant-connection.js
/**
 * @param {number[][]} edges
 * @return {number[]}
 * Time complexity: O(V + E), Since count(V) == count(E) in this case, tc becomes O(n)
 * Space complexity: O(n)
 */
var findRedundantConnection = function (edges) {
  const parents = Array.from({ length: edges.length + 1 }, (_, i) => i); // each node has itself as its parent
  const ranks = Array(edges.length + 1).fill(1); // every node starts off as rank 1

  // Return the root parent of a node,
  // while also doing path compressing at the same time
  function find(n) {
    if (n !== parents[n]) {
      parents[n] = find(parents[n]);
    }
    return parents[n];
  }

  // An uncomplete union algorithm for this specific problem
  // Return true if connection is successfully made (no cycle)
  // and vice versa
  function union(n1, n2) {
    const parent1 = find(n1);
    const parent2 = find(n2);

    // Cycle detected
    if (parent1 === parent2) {
      return false;
    }

    // No cycle -> connect nodes -> node with higher rank should be the parent
    if (ranks[parent1] > ranks[parent2]) {
      parents[parent2] = parent1;
      ranks[parent1] += ranks[parent2];
    } else {
      parents[parent1] = parent2;
      ranks[parent2] += ranks[parent1];
    }

    return true;
  }

  // Check each edge for cycle
  // The problem states that exactly 1 extra edge exist
  // So there's no need to handle the false case
  for (const [n1, n2] of edges) {
    if (!union(n1, n2)) {
      return [n1, n2];
    }
  }
};
